CIPs/0008.md

```diff
@@ -0,0 +1,74 @@
+# CIP 0008: Expand Metadata to general off-chain storage
+
+- Date: 2020-05-05
+- Author: @nambrot
+- Status: WIP
+
+## Overview
+
+In CIP3, the Celo protocol has started with a simple off-chain storage mechanism mainly focused on claims an account can make about their off-chain identity and communicate to other participants in the network. Over time, there have been additional use cases which have similar access patterns that prompted the creation of `Metadata` in the first place. I.e. data that does not require the expensive cost of "hard consensus", instead just needs to be authenticated by a single party.
+
+Examples include:
+- Transaction metadata such as comments or encrypted messaging in general
+- Profile information such as name or a profile picture
+- Unidirectional attestations like an audit or badges won in The Great Celo Stake Off
+- Signed statements/claims by a party for relase binaries/docker images/package hashes
+
+## Goals
+
+- Expand storage and indexing capabilities of this off-chain data
+- Encode the notion of decryption keys
+- Lightweight messaging system
+
+
+## Proposed Solution
+
+We propose to add another claim, called the `STORAGE` claim:
+```json
+{
+  "type": "STORAGE",
+  "timestamp": 1588886114907,
+  "address": "https://example.com/celo_storage"
+}
+```
+
+The address should be referred to as the root. The expectation here is that data can accessed via `root/${dataAddress}`. For private data, this should be encrypted to the recipient. Here are some common ways of modeling applications into this format:
+
+- For transaction metadata, the account can encrypt the metadata and put it in a file with the transaction hash as the metadata. Further CIPs could specify a standard for structured metadata.
+    - `dataAddress`: `transactionHash`
+    - file contents: Encrypted, structured transaction metadata (comments, or otherwise)
+- Data in relation to other accounts could be communicated to other accounts with:
+    - `dataAddress`: `otherAccountAddress`
+    - file contents: Structured data such as:
+        - `sessionKeys`: The session key(s) encrypted with dataEncryptionKey of `otherAccount` on `Accounts.sol`
+        - `attestations`: These could be signed data attestating things such an audit or badges
+- Profile information could be stored on this system with:
+    - `dataAddress`: `profileName` or `profilePicture`
+    - file contents: the data unencrypted, or encrypted to session keys available to the recipients
+- Releases:
+    - `dataAddress`: `hashOfRelease`
+    - file contents: The signature of the key on the hash of the release
+
+If no assocation between to the `dataAddress` should be leaked, the `dataAddress` itself could be encrypted. The recipient would then just have to decrypt all `dataAddress`s (which could be available in an index file).
+
+All file contents should always be signed by an authorized signer of the account to prevent tampering by the file hoster.
```

Comment:
Should we more concretely specify how the signature happens so there won't be compatibility issues?

Reply:
Good point. I was mostly thinking that each payload is along the lines of `{ data: any, signature: string }` where signature is over `JSON.stringify(data)`

Reply:
@timmoreton brought up that it would be nice if the files could be taken as is, and signatures are available under `{dataAddress}.signature`